onnxruntime/python/onnxruntime_cuda_temp_env.py

```diff
@@ -0,0 +1,61 @@
+import os
+import platform
+import re
+import site
+
+
+class TemporaryEnv:
+    def __init__(self, updates):
+        self.original_env = os.environ.copy()
+        os.environ.update(updates)
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        os.environ.clear()
+        os.environ.update(self.original_env)
+
+
+def get_nvidia_dll_paths() -> str:
+    # Get the site-packages path where nvidia packages are installed
+    site_packages_path = site.getsitepackages()[0]
+    nvidia_path = os.path.join(site_packages_path, "nvidia")
+
+    # Collect all directories under site-packages/nvidia that contain .dll files (for Windows)
+    dll_paths = []
+    for root, _, files in os.walk(nvidia_path):
+        if any(file.endswith(".dll") for file in files):
+            dll_paths.append(root)
+    return os.pathsep.join(dll_paths)
+
+
+def get_nvidia_so_paths() -> str:
+    # Get the site-packages path where nvidia packages are installed
+    site_packages_path = site.getsitepackages()[0]
+    nvidia_path = os.path.join(site_packages_path, "nvidia")
+
+    # Collect all directories under site-packages/nvidia that contain .so files (for Linux)
+    so_paths = []
+    # Regular expression to match `.so` optionally followed by `.` and digits
+    pattern = re.compile(r"\.so(\.\d+)?$")
+    for root, _, files in os.walk(nvidia_path):
+        for file in files:
+            if pattern.search(file):
+                so_paths.append(root)
+    return os.pathsep.join(so_paths)
+
+
+def setup_temp_env_for_ort_cuda():
+    # Determine platform and set up the environment accordingly
+    if platform.system() == "Windows":  # Windows
+        nvidia_dlls_path = get_nvidia_dll_paths()
+        if nvidia_dlls_path:
+            return TemporaryEnv({"PATH": nvidia_dlls_path + os.pathsep + os.environ.get("PATH")})
+        else:
+            return TemporaryEnv({"PATH": os.environ.get("PATH")})
```

Comment:
On Windows, you can do `os.add_dll_directory`.

Reply:
Here I am trying to use the TemporaryEnv for each inference session. At the end of each session, the PATH will roll back to the previous value.

Reply:
I suggest to check pytorch's preload logic as well (When `import torch`, cuda DLLs will be preloaded. Setting of PATH or LD_LIBRARY_PATH might not have effect when those cuda DLLs have been preloaded. Not sure there is a good way to verify whether preloaded DLLs are compatible):

https://github.com/pytorch/pytorch/blob/5f4a21dc58c7b0a732ae0dec8fdbf2dfbda4e7d5/torch/__init__.py#L155-L336

Reply:
Just updated to use `add_dll_directory` and `ctypes.CDLL` like @baijumeswani  suggested.

Reply:
But from the pytorch example, maybe we should move the dll/so loading to __init__.py